### PR TITLE
Split C++ into multiple files

### PR DESCRIPTION
**Pull request by halvfigur**
Split the relatively large iotea.cpp/hpp files in to smaller parts and
do the same for the relevant tests. This change has only affects the
file structure.

Also remove the examples "function_provider" and "event_consumer". These
examples where developed as a means to test the SDK and experiment with
the platform during development. The other examples cover the
functionality demonstrated by the removed examples.

Signed-off-by: Alexander Perlman <fixed-term.Alexander.Perlman@se.bosch.com>

### CODE REVIEW

**Comment by JochenKienzle** [comment]
I would recommend to add another reviewer because it is a very large pull request

File: src/sdk/cpp/lib/CMakeLists.txt [hunk 1]
```diff
@@ -38,8 +38,12 @@ add_library(iotea_sdk_cpp
   SHARED
     src/util.cpp
     src/logging.cpp
-    src/iotea.cpp
+    src/event.cpp
```
Comment (JochenKienzle) [inline src/sdk/cpp/lib/CMakeLists.txt]: Is there an order? If not we could introduce one (alphabetical) but this is only a 'nice to have'

File: src/sdk/cpp/lib/include/context.hpp [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+       context.Emit("temperature", degrees, "device");
+     * @endcode
+     *
+     * @tparam T The value to emit, must be either JSON or a value that can be
+     * implicitly converted to JSON.
+     * @param feature The name of the feature
+     * @param value The value payload of the event
+     * @param type The name of the type providing the feature
+     * @param instance The name of the instance
+     *
```
Comment (JochenKienzle) [inline src/sdk/cpp/lib/include/context.hpp]: unnecessary newline

File: src/sdk/cpp/lib/include/common.hpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+static constexpr char MQTT_TOPIC_NS[] = "iotea";
+static constexpr char DEFAULT_INSTANCE[] = "default";
+static constexpr char DEFAULT_TYPE[] = "default";
+
+using uuid_generator_func_ptr = std::function<std::string(void)>;
+
+}  // namespace core
+}  // namespace iotea
+
+#endif // SRC_SDK_CPP_LIB_INCLUDE_COMMON_HPP_
```
Comment (JochenKienzle) [inline src/sdk/cpp/lib/include/common.hpp:29]: An idea: All constants of the sdk could be defined here (in common place) and maybe renamed to constants?

**Review by JochenKienzle — CHANGES_REQUESTED** [review]
Looks good 👍 .
I have no real findings (only some very small findings/ideas). Please check

File: src/sdk/cpp/lib/include/common.hpp (hunk 3 above)
Comment (halvfigur) [reply src/sdk/cpp/lib/include/common.hpp:29]: I think it's better to have constants where they are used, it keeps the namespace clean.

File: src/sdk/cpp/lib/CMakeLists.txt (hunk 1 above)
Comment (halvfigur) [reply src/sdk/cpp/lib/CMakeLists.txt]: Re-ordred.

File: src/sdk/cpp/lib/include/context.hpp (hunk 2 above)
Comment (halvfigur) [reply src/sdk/cpp/lib/include/context.hpp]: Removed the unnecessary newline.

**Review by JochenKienzle — APPROVED** [review]
Looks good 👍

**Pull request merged**